internal/ackhandler/sent_packet_handler.go

```diff
@@ -503,7 +506,10 @@ func (h *sentPacketHandler) queueHandshakePacketsForRetransmission() error {
 }
 
 func (h *sentPacketHandler) queuePacketForRetransmission(p *Packet) error {
-	if _, err := h.packetHistory.QueuePacketForRetransmission(p.PacketNumber); err != nil {
+	if p.cannotBeRetransmitted {
+		return fmt.Errorf("sent packet handler BUG: packet %d already queued for retransmission", p.PacketNumber)
+	}
+	if err := h.packetHistory.MarkCannotBeRetransmitted(p.PacketNumber); err != nil {
```

Comment:
You could move stopRetransmissionsFor into MarkCannotBeRetransmitted and call that from here without changing behavior IIUC. That would decrease coupling between the SPH and history.

Reply:
It would work, but it would be really weird, since we're calling `MarkCannotBeRetransmitted` from `queuePacketForRetransmission` as well.